enarx-srv-traits/src/lib.rs

```diff
@@ -0,0 +1,120 @@
+//!
+//! Once a `Preparer` is obtained, the keep manager will call `Loader::load()`
+//! to load all the pages of the static binary. Then finally, the manager
+//! will call `Preparer::next()` to specify the entry point and conclude the
+//! preparation state.
+//!
+//! ## Building
+//!
+//! The purpose of the building state is to actually create a full keep
+//! with all the pages actually loaded. The workflow is similar to the
+//! preparation state.
+//!
+//! ## Attestation
+//!
+//! The purpose of the attestation state is attest the keep to a remote
+//! entity (usually the tenant). This state proxies packets back and forth
+//! between the keep and the tenant. Note that the contents of attestation
+//! packets is defined by the specific technology in question.
+//!
+//! ## Launching
+//!
+//! The purpose of the launching state is to proxy communication between
+//! the tenant and the keep after attestation. Note that the intent of this
+//! state is to proxy integrity protected ciphertext. The manager should
+//! have no visibility into its contents.
+//!
+//! ## Running
+//!
+//! The purpose of the running state is to handle syscalls from the keep.
+
+use std::convert::TryFrom;
+use std::io::{Read, Result, Write};
+use std::num::TryFromIntError;
+use std::ops::BitAnd;
+
+use bitflags::bitflags;
+
+bitflags! {
+    /// Memory access permissions
+    ///
+    /// Raw numeric values are from the ELF specification.
+    pub struct Access: u8 {
+        /// Execute access
+        const X = 1 << 0;
+
+        /// Write access
+        const W = 1 << 1;
+
+        /// Read access
+        const R = 1 << 2;
+    }
+}
+
+/// A keep request
+pub enum Request {
+    Halt,
+}
+
+/// A keep reply
+pub enum Reply {
+    Halt,
+}
+
+/// A trait for supplying statically linked pages into the keep
+pub trait Loader {
+    fn load(&mut self, pages: &[u8], offset: u64, access: Access) -> Result<()>;
+}
+
+/// Prepares to build the keep
+pub trait Preparer: Sized + Loader {
+    type Builder: Builder;
+
+    /// Creates a new Preparer
+    ///
+    /// Returns None when this preparer type is not supported.
+    fn new() -> Result<Option<Self>>;
+
+    /// Completes the preparation process after loading all pages
+    ///
+    /// This specifies the entry point for the binary and creates the builder.
+    fn next(self, entry: u64) -> Result<Self::Builder>;
+}
+
+/// Builds a keep
+pub trait Builder: Sized + Loader {
+    type Keep: Keep;
+
+    /// Completes the building process after loading all pages
+    ///
+    /// This specifies the entry point for the binary and the socket on which
+    /// to perform attestation and encrypted communication. Upon success,
+    /// returns the new keep.
+    fn next(self, entry: u64, socket: impl Read + Write) -> Result<Self::Keep>;
+}
+
+/// Executes a keep
+pub trait Keep {
+    /// Runs the keep until a event occurs.
+    fn run(&mut self, reply: Reply) -> Result<Reply>;
```

Comment:
This area still needs some thinking. But the basic idea is this. The keep runs until it needs something from the keep manager. This is a `Request`. The keep manager handles the `Request` and sends the `Reply` back to the keep.

I haven't solved the problem that the first time you call `run()` you don't have a `Reply` yet because there has been no request. Suggestions welcome.

Reply:
Ah okay, thanks for the explanation! It wouldn't work as an `Option<Reply>`?

Reply:
I'll probably do that for now. But I'd like the compiler to enforce a `Reply` after the first `Request` is received.

Reply:
We could make a special `Reply` that is used only for the first run of the Keep. A kind of stub. In that case, we may need to keep some state to indicate we have seen it, as additional "first run" replies will probably be considered spurious and discarded since the Keep has already been running.

The vocabulary here is probably what we're getting caught up on for that first case. We could enclose a `Reply` or a `Request` in a top-level type called `Message`.

Reply:
I have a solution for this coming in a forthcoming push. My solution not only ensures the correct value as input for the first call, but it also ensures that the compiler enforces the correct `Reply` data corresponds to the `Request`.